CIP-0175? | Stake Pool Hot Credentials (#1140)
cardano-foundation/CIPs: CIP-0175/README.md

remain offline.

Authorization for consensus-critical voting must be ledger-visible and
ledger-validated. Existing Calidus keys from CIP-0151 rely on transaction
metadata, but is not the right substrate for on-chain voting.
A certificate-based design provides explicit state transitions,
deterministic validation, and consistent tooling semantics.

This CIP enables day-to-day governance operation through authorized hot
credentials while preserving cold-key voting as it already works today.

## Specification

### Terminology

- **Pool cold key**: The Ed25519 key whose hash defines the Pool ID.
- **Pool ID**: The blake2b-224 hash of the pool cold verification key.
- **Hot credential**: A governance `Credential` authorized for a pool's
  governance voting role. It may be either a key hash credential or script hash
  credential.
- **Hot authorization map**: Ledger state map
  `poolGovHotCreds : Map PoolId Credential`.
- **Vote source**: Authorization source for a recorded SPO vote:
  `Cold | Hot`.

### Certificate Semantics

This CIP introduces two new stake-pool governance certificates:

- `AuthStakePoolHotKey(poolId, hotCred)`

> - `ResignStakePoolHotKey(poolId)`

Reviewer: We can get away with adding only one certificate instead:
```haskell
AssignStakePoolHotCredential (KeyHash StakePool) (Maybe (Credential StakePoolHot))
```
or in pseudocode:
```
- `AssignStakePoolHotCredential(poolId, hotCred)`
```

* `Nothing` case would remove hoot credential authorization if there was any
* while `Just` case would overwrite whatever was there before with a new hot credential

The main benefit of this approach would be reduction in complexity (one certificate vs two) and reduction in necessary validation and contention prevention. With two certificates we would probably have to add some validation against prior state. For example in case of resign we would probably have to check that there was a hot credential there before.

In any case. Simpler is better, IMHO.

We'll also need to describe this as necessary change to CDDL

---
CIP: 175
Title: Stake Pool Hot Credentials
Category: Ledger
Status: Proposed
Authors:
    - Ryan Wiley <rian222@gmail.com>
Implementors: []
Discussions:
    - CIP-0151: https://github.com/cardano-foundation/CIPs/pull/999
    - Original PR: https://github.com/cardano-foundation/CIPs/pull/1140
Created: 2026-01-22
License: CC-BY-4.0
---

## Abstract

Stake Pool Operators (SPOs) currently sign on-chain governance votes with pool
cold keys. This CIP introduces a new on-chain certificate that lets a pool cold
key authorize a hot credential for a stake pool. For
`StakePoolVoter poolId`, the ledger accepts either:

- the pool cold key witness, or

> - a witness that satisfies the currently authorized hot credential.

Reviewer: I'd suggest doing the same thing as with CC members, only cold key can authorize new hot credentials. If we don't do that it will require a bit of work:
* Support for Plutus scripts will need to be implemented for this new certifying approach
* Certificate will need to be explicit which hot or cold is to be used for authorization, since "either or" will not be automagically inferred by the ledger

- **Pool ID**: The blake2b-224 hash of the pool cold verification key.
- **Calidus key**: The Ed25519 public key registered in CIP-0151 metadata for a
  stake pool (registration payload field 7).
- **Active Calidus key**: The Calidus key from the valid CIP-0151 registration
  with the highest nonce for a given pool. A 32-byte zero key indicates
  revocation (no active key).

### Data Source: CIP-0151 Registration Metadata

The ledger recognizes CIP-0151 registrations recorded under metadata label
`867`, version `2`, with **Scope** = stake pool (`[1, poolID]`). The following
fields are required (as per CIP-0151):

- Version (0) = 2
- Registration Payload (1)
- Registration Witness (2)

Within the Registration Payload, the following fields are required:

- Scope (1)
- Feature Set (2)
- Validation Method (3)
- Nonce (4)

The optional Calidus Key (7) is used as the candidate key to authorize SPO
voting.

### Validation of CIP-0151 Registrations (Ledger Rule Additions)

To prevent unauthorized key claims, the ledger MUST validate CIP-0151
registrations before they can influence voting authorization. A registration is
**valid** for ledger purposes when all of the following hold:

> 1. The metadata conforms to CIP-0151 version 2 and scope = stake pool.

Reviewer 2: solved by changing approach as indicated in https://github.com/cardano-foundation/CIPs/pull/1140#issuecomment-3947800536

---
CIP: 175
Title: Stake Pool Hot Credentials
Category: Ledger
Status: Proposed
Authors:
    - Ryan Wiley <rian222@gmail.com>
Implementors: []
Discussions:
    - CIP-0151: https://github.com/cardano-foundation/CIPs/pull/999

>     - Original PR: https://github.com/cardano-foundation/CIPs/pull/1140

Reviewer 2: ```suggestion
    - Original PR: https://github.com/cardano-foundation/CIPs/pull/1140
    - CIP-0151 PR: https://github.com/cardano-foundation/CIPs/pull/999
```
Merged in the meantime; though PR discussion likely more relevant than CIP-0151 content itself; while still not more relevant than `this` PR (therefore appearing first).

Reviewer 2: @Cerkoryn I'm committing this to get an immediate confirmation that this passes CIP validation (or display why it doesn't).

  If none exists, this is a valid no-op.

### Certificate CDDL

```cddl
; Numeric certificate tag allocation is TBD in Ledger CDDL integration.
assign_stake_pool_hot_credential_tag = uint

assign_stake_pool_hot_credential_cert = [
  assign_stake_pool_hot_credential_tag,
  pool_id : pool_keyhash,
  hot_credential : null / credential
]
```

### Certificate Validation Rules

For `AssignStakePoolHotCredential(poolId, hotCredOpt)`:

1. `poolId` MUST identify a currently registered pool.
2. The transaction MUST include a valid witness by that pool's cold key.
3. If multiple `AssignStakePoolHotCredential` certificates for the same
   `poolId` appear in one transaction, they are processed in transaction order;
   the last one determines the final hot credential state for that `poolId`.

### Ledger State Extension

Introduce:

```
poolGovHotCreds : Map PoolId Credential
```

State transitions:

- `AssignStakePoolHotCredential(poolId, hotCredOpt)` with
  `hotCredOpt = hotCred` sets `poolGovHotCreds[poolId] = hotCred`
  (overwrite allowed).
- `AssignStakePoolHotCredential(poolId, hotCredOpt)` with
  `hotCredOpt = null` removes `poolId` from the map if present (otherwise

>   no-op).

Reviewer: We need to explicitly list here that, unlike regular pool params update, this certificate will cause the credential to be updates for the stake pool immediately after transaction is applied on chain. Which is also the reason why it had to be implemented as a new certificate, instead of relying on the usual RegPool certificate that is capable of updating stake pool state.

No uniqueness constraint is imposed on `hotCred`. The same hot credential MAY be
authorized for multiple pools.

Pool retirement lifecycle:

- If a pool has a retirement certificate scheduled but not yet enacted, its hot
  credential remains valid.
- When retirement is enacted, the ledger clears that pool's

>   `poolGovHotCreds` entry.

Reviewer: Technically speaking it is not needed, since upon retirement the whole stake pool state is cleared out, which will include the hot credential.
But, it's ok to keep it here, since what I mentioned above is just an implementation detail (i.e. it would be impossible to retain the hot credential for a retired pool)

### SPO Voter Representation and Authorization

This CIP changes the SPO voter representation to:

```haskell
StakePoolVoter !(KeyHash 'StakePool) (Maybe (Credential Hot))
```

Authorization semantics:

1. `StakePoolVoter poolId Nothing` is a cold-authorized SPO vote. The
   transaction MUST include a valid witness by the pool cold key for `poolId`.
2. `StakePoolVoter poolId (Just hotCred)` is a hot-authorized SPO vote. The
   transaction MUST include witnesses that satisfy `hotCred`, and the ledger
   MUST verify that `poolGovHotCreds[poolId] = hotCred`.

Hot credential satisfaction rules:

- If `hotCred` is a key credential, the corresponding vkey witness is required.
- If `hotCred` is a script credential, a valid governance voting script witness
  is required under existing governance voting-script rules for
  `StakePoolVoter poolId (Just hotCred)`.
- Both native scripts and Plutus scripts are supported.

> 

Reviewer: ```suggestion
- Existing `Voting` script purpose will be used for Plutus script execution, since it was unused for stake pools before

```

All other vote semantics (vote options, anchors, role structure, timing windows,
and tallying model) remain as defined by CIP-1694 unless explicitly modified by
this CIP.

### Vote Type and Overwrite Rules

For each active `(govActionId, poolId)` vote slot, the ledger records only the
latest SPO vote together with its `VoteSource` (`Cold` or `Hot`). Historical
overwritten SPO votes do not need to be retained.

Overwrite rules:

1. A `Cold` vote MAY replace the current vote (`Cold` or `Hot`) for the same
   `(govActionId, poolId)`.
2. A `Hot` vote MAY replace the current `Hot` vote for the same
   `(govActionId, poolId)`.
3. A `Hot` vote MUST NOT replace the current `Cold` vote for the same

>    `(govActionId, poolId)`.

Reviewer: I am suggesting "SHOULD", because ignoring such a vote does not compromise the system, but would result in potentially dangerous bugs in tooling and confusing users.
```suggestion
3. A `Hot` vote MUST NOT replace the current `Cold` vote for the same
   `(govActionId, poolId)`. An attempt to do so SHOULD result in a predicate failure, instead of being silently ignored.
```

Effectively, once any cold-authorized vote exists for a given
`(govActionId, poolId)`, that pair is locked against hot-authorized overwrites.

### Authorization-Change Vote Invalidation

> If `AssignStakePoolHotCredential(poolId, hotCredOpt)` results in an actual
> authorization-state change for `poolId`, the ledger MUST clear the current SPO
> votes for that `poolId` across still-active governance actions whose recorded
> `VoteSource` is `Hot`.

Reviewer: It is not necessary to clear out the vote. In fact we do not want to do this, because that would be an expensive operation. It is much simpler to not count votes during ratification that were made by a hot credential that does not match the up to date vote in the stake pool state.
```suggestion
If `AssignStakePoolHotCredential(poolId, hotCredOpt)` results in an actual
authorization-state change for `poolId`, the ledger MUST not use this current SPO
votes for that `poolId` across still-active governance actions whose recorded
`VoteSource` is `Hot`.
```

If `AssignStakePoolHotCredential(poolId, hotCredOpt)` does not change the
resulting authorization state for `poolId`, no vote invalidation occurs.

Current `Cold` votes remain valid.

### Ledger Rule Integration

This CIP integrates at existing governance ledger rule boundaries:

- **`UTXO`** applies certificate-driven updates to `poolGovHotCreds` and
  enforces transaction-level constraints for this certificate type.
- **`UTXOW`** validates the explicit SPO voter form, requiring either the pool
  cold witness for `StakePoolVoter poolId Nothing` or witnesses satisfying
  `hotCred` plus `poolGovHotCreds[poolId] = hotCred` for
  `StakePoolVoter poolId (Just hotCred)`.
- **`GOV`** (vote state handling) records `VoteSource` for stake-pool votes and
  stores only the latest current vote per `(govActionId, poolId)`, applying
  cold/hot overwrite rules plus hot-vote invalidation on actual
  authorization-state changes.

The proposal does not introduce a new governance role. It extends
`StakePoolVoter` so the SPO vote explicitly carries whether it is using cold

> authorization or a specific hot credential.

Reviewer: This whole section goes into how this CIP would be implemented. I'd recommend avoiding these details, especially since they are inaccurate.
```suggestion
```

- **Simplicity**: A single certificate with nullable payload reduces complexity
  in specification, validation, and tooling, and the ledger only needs the
  latest vote and its type.

### Backward Compatibility

- Existing cold-key SPO voting remains valid through
  `StakePoolVoter poolId Nothing`.
- This CIP extends the SPO voter payload to include
  `Maybe (Credential Hot)` explicitly.
- This CIP introduces no dependency on transaction metadata.
- CIP-0151 remains compatible as an off-chain identity/authentication mechanism.
  Operators MAY reuse the same underlying key material for both systems, but it
  is not required by consensus.

### Security Considerations

- **Hot credential compromise near deadline**:
  If hot keys are suspected to be compromised, changing or revoking the
  authorized hot credential immediately invalidates current hot-authorized SPO
  votes for that pool on still-active governance actions, while current
  cold-authorized votes remain valid. Operators can then recast under current
  authorization (including with the cold key). Additionally, large stake pool
  groups can mitigate this risk further by choosing to use a multisig hot

>   credential.

Reviewer: It would probably be useful to note here that during one epoch delay before gov action enactment, neither hot credential authorization change, not a vote change can affect that enactment. This is also applicable to votes by DReps and CC members alike.

- **Blast radius for shared hot credentials across multiple pools**:
  This is allowed by design. Operators should consider script credentials (for
  example multisig and timelock designs) to reduce single-key compromise risk.
- **Intentional delegation via shared or third-party hot credentials**:
  A pool cold key can authorize any `Credential`, including one controlled by a
  different operator or reused across multiple pools. Votes cast as
  `StakePoolVoter poolId (Just hotCred)` then count for each pool that has
  authorized that `hotCred`. This representative behavior is an intentional
  capability, not a validation flaw, but it can concentrate SPO voting
  influence. If that outcome is undesirable, pool operators can revoke or
  replace the hot credential and, for the current governance action, reassert

>   intent with a cold-authorized vote.

Reviewer: This is incorrect. We should not apply automatically one vote made by one hot credential for all stake pools that authorized the same hot credential. In fact this is not what we discussed during the working group either.

There are technical reasons why we do not want to implement it this was. But also, we do not want one SPO A mimicking votes of another SPO B, without authorization, by the SPO B.

Therefore, I'd recommend changing this text to something like this.
```suggestion
- **Intentional delegation via shared or third-party hot credentials**:
  A pool cold key can authorize any `Credential`, including one controlled by a
  different operator or reused across multiple pools. Many votes with different
  `poolId`s, but the same `hotPoolCred` can be submitted like this: 
  `StakePoolVoter poolId (Just hotPoolCred)`. In fact as many as there are pools 
  that authorized that `hotPoolCred`. Note that there is no requirement for all 
  those votes to be the same, eg. for some pools it could be Yes, for others No.
  This representative behavior is an intentional capability, not a validation
  flaw, but it can concentrate SPO voting influence. If that outcome is
  undesirable, pool operators can revoke/replace the hot credential or replace
  the vote by hot credential for the current governance action with a 
  coldkey-authorized vote.
```
